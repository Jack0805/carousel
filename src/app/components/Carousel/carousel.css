.container {
    width: 100%;
    padding:20px;
}

.carousel {
    display: flex;
    justify-content: center;
    overflow-x: hidden;
    width: 100%;
}

.item-image {
    width: 260px;
}