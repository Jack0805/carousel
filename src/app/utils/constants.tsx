export const endPoint = '/data/data.json'