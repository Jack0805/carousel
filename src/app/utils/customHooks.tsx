'use client'
import {useState, useEffect, useRef} from 'react';

export function useViewportWidth() {
    const isBrowser = typeof window !== 'undefined';
    const [viewportWidth, setViewportWidth] = useState(isBrowser ? window.innerWidth : 0);

    useEffect(() => {

        if (!isBrowser) {
            return;
        }
        const handleResize = () => {
            setViewportWidth(window.innerWidth);
        };

        window.addEventListener('resize', handleResize);
        return () => {
            window.removeEventListener('resize', handleResize);
        };
    }, [isBrowser]);

    switch (true) {
        case viewportWidth > 1600:
            return 6;
        case viewportWidth > 1080:
            return 5;
        case viewportWidth > 720:
            return 4;
        default:
            return 3;
    }
}

interface UseAutoFocusState {
    index:number,
}
export function useAutoFocus({index}: UseAutoFocusState) {

    const myDivRef = useRef<HTMLDivElement>(null);

    useEffect(() => {

        if (myDivRef.current && index == 0) {
            myDivRef.current.focus();
        }
        const handleClickOutside = (event: MouseEvent) => {
            if (myDivRef.current && !myDivRef.current.contains(event.target as Node) && index == 0) {
                myDivRef.current.focus();
            }
        };

        document.addEventListener('click', handleClickOutside);

        return () => {
            document.removeEventListener('click', handleClickOutside);
        };

    }, []);

    return {
        myDivRef
    }
}