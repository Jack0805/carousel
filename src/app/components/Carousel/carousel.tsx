import React, {useEffect, useState, KeyboardEvent} from 'react';
import {
    fetchMoviesThunk,
    getNextMoviesOnScreenArr,
    getPreviousMoviesOnScreenArr,
    setCurrentMovieData,
    MovieState
} from "@/app/state/movies/movieSlice";
import {RootState} from "@/app/state/store";
import {useAppDispatch,useAppSelector} from "@/app/state/hooks";
import Movie from './movie'
import { useRouter } from "next/navigation";
import './carousel.css'
export interface CarouselProps {
    endPoint: string
    maxNumberOfItem: number
}
export interface OnClickProps {
    movie:  MovieState,
    selectedIndex: number
}
const Carousel = ({endPoint,maxNumberOfItem}:CarouselProps) => {
    const previousPageState = useAppSelector((state: RootState) => state.movies.previousPageState)
    const [selectedIndex, setSelectedIndex] = useState(previousPageState.selectedIndex)
    const moviesOnScreen = useAppSelector((state: RootState) => state.movies.moviesOnScreen);
    const currentPageNumber = useAppSelector((state: RootState) => state.movies.currentPage);
    const numOfPages = useAppSelector((state: RootState) => state.movies.numOfPages);
    const numbOfLastPageMovies = useAppSelector((state: RootState) => state.movies.numbOfLastPageMovies);
    const dispatch = useAppDispatch();
    const router = useRouter();

    useEffect(() => {
        if (previousPageState.pageNum == 1 || currentPageNumber == 1 ) {
            dispatch(fetchMoviesThunk({endPoint, maxNumberOfItem, errorMessage: 'An unknown error occurred, please try again later'}))
        }
    }, [maxNumberOfItem]);
    const handleKeyPress = (event: KeyboardEvent<HTMLDivElement>) => {
        if (event.key === 'ArrowLeft') {
            if (selectedIndex == 0) {
                if (currentPageNumber != 1) {
                    dispatch(getPreviousMoviesOnScreenArr({currentPageNumber, maxNumberOfItem}))
                    setSelectedIndex(maxNumberOfItem -1)
                }
            } else {
                setSelectedIndex(selectedIndex - 1)
            }

        } else if (event.key === 'ArrowRight') {
            if (selectedIndex < maxNumberOfItem - 1) {
                if (currentPageNumber == numOfPages && selectedIndex == numbOfLastPageMovies - 1){


                } else {
                    setSelectedIndex(selectedIndex + 1)
                }
            } else {
                if (currentPageNumber != numOfPages){
                    dispatch(getNextMoviesOnScreenArr({currentPageNumber, maxNumberOfItem}))
                    setSelectedIndex(0)
                }
            }
        } else if (event.key === 'Enter'){
            router.push(`/program/?mid=${moviesOnScreen[selectedIndex].id}&selectedIndex=${selectedIndex}&page=${currentPageNumber}`)
        }
    }

    const handleOnClick = (props:OnClickProps) => {
        setSelectedIndex(props.selectedIndex)
        dispatch(setCurrentMovieData({...props.movie}))
    }

    return (
        <div className="container">
            <div className="carousel">
                {
                    moviesOnScreen.map((movie,idx) => {
                        return (
                            <Movie
                                index={idx}
                                movie={movie}
                                key={movie.id}
                                selected={idx===selectedIndex}
                                selectedIndex={idx}
                                currentPage = {currentPageNumber}
                                handleClick = {handleOnClick}
                                handleKeyPress={handleKeyPress}/>

                        )
                    })
                }
            </div>
        </div>
    );
};

export default Carousel;