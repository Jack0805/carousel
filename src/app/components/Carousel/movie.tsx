import React, {KeyboardEvent} from 'react';
import {MovieState, setCurrentMovieData} from '../../state/movies/movieSlice'
import withBorder from '../../utils/withborder'
import Link from "next/link";
import {useAutoFocus} from '../../utils/customHooks'
import {OnClickProps} from './carousel'

export interface movieProps {
    movie: MovieState,
    selected: boolean,
    selectedIndex: number,
    handleClick: (props: OnClickProps) => void | null;
    handleKeyPress: (event: KeyboardEvent<HTMLDivElement>) => void;
    index: number | null,
    currentPage: number | null
}
const Movie = (props: movieProps) => {
    const {id, image} = props.movie
    const {myDivRef} = useAutoFocus({index:props.index || 0})
    return (
            <div onClick={() => props.handleClick({movie: props.movie, selectedIndex:props.selectedIndex})} onKeyDown={(e) => props.handleKeyPress(e)} tabIndex={0} ref={myDivRef} style={{ outline: 'none' }}>
                <Link href={{ pathname: '/program', query: { mid:id, selectedIndex: props.selectedIndex, page: props.currentPage, from: 'home' }}}><img className="item-image" src={image}></img></Link>
            </div>
    );
};
export default withBorder(Movie);