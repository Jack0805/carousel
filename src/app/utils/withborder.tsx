import React, {ComponentType, useRef, useEffect} from 'react';
import {movieProps} from "@/app/components/Carousel/movie";

const withBorder = <P extends movieProps>(Component: ComponentType<P>) => {
    return (props: P)  => (
        <div
            className={props.selected ? 'image-container-with-border':'image-container-without-border'}
        >
            <Component {...props}/>
        </div>
    );
};

export default withBorder;