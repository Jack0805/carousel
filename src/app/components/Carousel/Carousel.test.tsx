import { render } from '@testing-library/react';
import React from 'react';
import Carousel from "@/app/components/Carousel/carousel";
import {endPoint} from "@/app/utils/constants";
import {Provider} from "react-redux";
import configureMockStore from 'redux-mock-store';
import '@testing-library/jest-dom';
import mockFetch from '../../utils/jest';
import thunk from 'redux-thunk';

jest.mock('next/navigation', () => ({
    useRouter: jest.fn()
}))
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Movies Component', () => {


    it('should render correctly', async () => {
        const mockData = await mockFetch.get()

        const store = mockStore({
            movies: {
                movies: mockData,
                moviesOnScreen: mockData.slice(1,6),
                previousPageState: {
                    selectedIndex: 0,
                    pageNum: 1
                }
            },
        });
        const { container } = render(
            <Provider store={store}>
                <Carousel endPoint={endPoint} maxNumberOfItem={6} />
            </Provider>,
        );
        expect(container.firstChild).toMatchSnapshot();
    });
})
