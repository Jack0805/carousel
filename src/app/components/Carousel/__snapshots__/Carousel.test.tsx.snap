// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies Component should render correctly 1`] = `
<div
  class="container"
>
  <div
    class="carousel"
  >
    <div
      class="image-container-with-border"
    >
      <div
        style="outline: none;"
        tabindex="0"
      >
        <a
          href="/program?mid=65737&selectedIndex=0&page=&from=home"
        >
          <img
            class="item-image"
            src="https://streamcoimg-a.akamaihd.net/000/657/37/65737-PosterArt-15bee119eb92a5f33670e6bd3f1af967.jpg?resize=512px:*&quality=75&preferredFormat=image/jpeg"
          />
        </a>
      </div>
    </div>
    <div
      class="image-container-without-border"
    >
      <div
        style="outline: none;"
        tabindex="0"
      >
        <a
          href="/program?mid=67517&selectedIndex=1&page=&from=home"
        >
          <img
            class="item-image"
            src="https://streamcoimg-a.akamaihd.net/000/675/17/67517-PosterArt-d740bf2159cf9c3eb1b22db4043076ff.jpg?resize=512px:*&quality=75&preferredFormat=image/jpeg"
          />
        </a>
      </div>
    </div>
    <div
      class="image-container-without-border"
    >
      <div
        style="outline: none;"
        tabindex="0"
      >
        <a
          href="/program?mid=56197&selectedIndex=2&page=&from=home"
        >
          <img
            class="item-image"
            src="https://streamcoimg-a.akamaihd.net/000/561/97/56197-PosterArt-469c02a0132ce2629859f70303083c85.jpg?resize=512px:*&quality=75&preferredFormat=image/jpeg"
          />
        </a>
      </div>
    </div>
    <div
      class="image-container-without-border"
    >
      <div
        style="outline: none;"
        tabindex="0"
      >
        <a
          href="/program?mid=26702&selectedIndex=3&page=&from=home"
        >
          <img
            class="item-image"
            src="https://streamcoimg-a.akamaihd.net/000/267/02/26702-PosterArt-fa4764f2e5af823683130cec70f227ae.jpg?resize=512px:*&quality=75&preferredFormat=image/jpeg"
          />
        </a>
      </div>
    </div>
    <div
      class="image-container-without-border"
    >
      <div
        style="outline: none;"
        tabindex="0"
      >
        <a
          href="/program?mid=3019990&selectedIndex=4&page=&from=home"
        >
          <img
            class="item-image"
            src="https://streamcoimg-a.akamaihd.net/000/301/9990/3019990-PosterArt-5984e294f596ed0ad0c006162770e080.jpg?resize=512px:*&quality=75&preferredFormat=image/jpeg"
          />
        </a>
      </div>
    </div>
  </div>
</div>
`;
